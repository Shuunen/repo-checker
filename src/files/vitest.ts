import { FileBase } from '../file.ts'

// eslint-disable-next-line no-export
export class VitestFile extends FileBase {
  // eslint-disable-next-line jsdoc/require-jsdoc
  // eslint-disable-next-line max-lines-per-function
  public async start() {
    if (!this.data.isUsingVitest) return
    const hasMtsFile = await this.fileExists('vitest.config.mts')
    const hasJsFile = await this.fileExists('vitest.config.js')
    const hasTsFile = await this.fileExists('vitest.config.ts')
    const hasMjsFile = await this.fileExists('vitest.config.mjs')
    if (hasMtsFile) await this.inspectFile('vitest.config.mts')
    /* c8 ignore next 3 */
    if (hasJsFile) await this.inspectFile('vitest.config.js')
    if (hasTsFile) await this.inspectFile('vitest.config.ts')
    if (hasMjsFile) await this.inspectFile('vitest.config.mjs')
    this.couldContains('a pool thread setting', /pool: 'threads'/u)
  }
}
