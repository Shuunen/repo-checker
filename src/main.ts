/* eslint-disable unicorn/no-process-exit */
import arg from 'arg'
import { existsSync, writeFileSync } from 'fs'
import requireFromString from 'require-from-string'
import { version } from '../package.json'
import { check } from './check'
import { dataDefaults, dataFileHomePath, dataFileName, home, ProjectData, repoCheckerPath, templatePath } from './constants'
import { log } from './logger'
import { join, readFileInFolder, resolve } from './utils'

async function initDataFile (doForce = false): Promise<void> {
  log.line()
  const fileExists = existsSync(dataFileHomePath)
  if (fileExists && !doForce) {
    log.warn('repo-checker data file', dataFileHomePath, 'already exists, use --force to overwrite it')
    return
  }
  const fileContent = await readFileInFolder(templatePath, dataFileName)
  const filePath = join(home, dataFileName)
  writeFileSync(filePath, fileContent)
  if (existsSync(filePath)) log.info('repo-checker data file successfully init, you should edit :', dataFileHomePath)
  else log.error('repo-checker failed at creating this file :', dataFileHomePath)
}

async function getData (argument = '', target = ''): Promise<ProjectData> {
  const dataPath = await getDataPath(argument, target)
  if (dataPath.length === 0) return dataDefaults
  log.info('loading data from', dataPath)
  return requireFromString(await readFileInFolder(dataPath, ''))
}

async function getDataPath (argument = '', target = ''): Promise<string> {
  const fileExists = existsSync(join(repoCheckerPath, argument))
  if (argument.length > 0 && fileExists) return join(repoCheckerPath, argument)
  const dataFileTargetPath = join(target, dataFileName)
  if (existsSync(dataFileTargetPath)) return dataFileTargetPath
  if (existsSync(dataFileHomePath)) return dataFileHomePath
  log.warn('you should use --init to prepare a data file to enhance fix')
  log.info('because no custom data file has been found, default data will be used')
  return ''
}

function getTarget (argument = ''): string {
  if (argument.length > 0) return resolve(argument)
  log.line()
  log.info('no target specified via : --target=path/to/directory')
  log.info('targeting current directory...')
  return process.cwd()
}

export async function start (): Promise<void> {
  const options = arg({ '--init': Boolean, '--force': Boolean, '--target': String, '--data': String, '--fix': Boolean, '--quiet': Boolean, '--version': Boolean, '-v': Boolean }, { argv: process.argv.slice(2) })
  if ((options['--version'] ?? false) || (options['-v'] ?? false)) {
    console.log(version)
    process.exit(0)
  }
  const doForce = options['--force']
  if (options['--init'] ?? false) {
    await initDataFile(doForce).catch(error => log.error(error))
    return
  }
  const doFix = options['--fix']
  const quiet = options['--quiet'] ?? false
  log.noConsole = quiet
  const target = getTarget(options['--target'])
  const data = await getData(options['--data'], target)
  data.quiet = data.quiet || quiet
  log.start(doFix)
  await check(target, data, doFix, doForce)
}
