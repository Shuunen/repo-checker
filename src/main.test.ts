import { clone } from 'shuutils'
import { expect, it } from 'vitest'
import { defaultOptions, getData, getFlags, getOptions, initDataFile, start } from './main'

function cleanTargetForSnap(options: Readonly<ReturnType<typeof getOptions>>) {
  const clean = clone<Partial<typeof options>>(options)
  // we need to get rid of targets like c:\Users\MyUser or /home/another-variable/ because they are not the same on every machine
  // biome-ignore lint/performance/noDelete: let me clean the snapshot ^^'
  delete clean.target
  return clean
}

it('parseOptions A defaults', () => {
  expect(cleanTargetForSnap(getOptions({}))).toMatchSnapshot()
})

it('parseOptions B non existing target', () => {
  expect(cleanTargetForSnap(getOptions({ '--fix': true, '--target': 'my-folder' }))).toMatchSnapshot()
})

it('parseOptions C show help', () => {
  expect(cleanTargetForSnap(getOptions({ '--help': true }))).toMatchSnapshot()
})

it('getFlags A', () => {
  expect(getFlags()).toMatchInlineSnapshot(`
    {
      "_": [],
    }
  `)
})

it('start A defaults', () => {
  // we dont await here because there no need to pollute the get the whole check result here, see check.test.ts
  expect(start()).toMatchInlineSnapshot('Promise {}')
})

it('start B show help', async () => {
  const options = { ...defaultOptions, willShowHelp: true }
  expect(await start(options)).toMatchInlineSnapshot(`
    {
      "failed": [],
      "passed": [
        "show-help",
      ],
      "warnings": [],
    }
  `)
})

it('start C show version', async () => {
  const options = { ...defaultOptions, willShowVersion: true }
  expect(await start(options)).toMatchInlineSnapshot(`
    {
      "failed": [],
      "passed": [
        "show-version",
      ],
      "warnings": [],
    }
  `)
})

it('start D init', () => {
  const options = { ...defaultOptions, willInit: true }
  expect(start(options)).toMatchInlineSnapshot('Promise {}')
})

it('start E verbose', () => {
  const options = { ...defaultOptions, isVerbose: true }
  expect(start(options)).toMatchInlineSnapshot('Promise {}')
})

it('start F quiet', () => {
  const options = { ...defaultOptions, isQuiet: true }
  expect(start(options)).toMatchInlineSnapshot('Promise {}')
})

it('start G fix', () => {
  const options = { ...defaultOptions, canFix: true }
  expect(start(options)).toMatchInlineSnapshot('Promise {}')
})

it('getData A empty target : current .repo-checker.json', async () => {
  expect(await getData()).toMatchInlineSnapshot(`
    {
      "maxSizeKo": 50,
    }
  `)
})

it('getData B non-existing target : default data', () => {
  // we dont await here because there no need to pollute the get the whole default data here, see utils.test.ts
  expect(getData('unknown')).toMatchInlineSnapshot('Promise {}')
})

it('getData C mal-formatted json target : default data', async () => {
  // we dont toThrowErrorMatchingSnapshot because the JSON parse error message is localized -.-''
  await expect(async () => await getData('src/mocks/tsProject/sub-folder')).rejects.toThrowError()
})

it('initDataFile A default to current folder but data file already exists', async () => {
  expect(await initDataFile()).toMatchInlineSnapshot(`
    {
      "failed": [],
      "passed": [],
      "warnings": [
        "data-file-already-exists",
      ],
    }
  `)
})

it('initDataFile B in a temp folder', async () => {
  expect(await initDataFile('node_modules', true)).toMatchInlineSnapshot(`
    {
      "failed": [],
      "passed": [
        "init-data-file",
      ],
      "warnings": [],
    }
  `)
})
