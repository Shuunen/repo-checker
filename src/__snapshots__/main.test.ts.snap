// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getOptions A defaults 1`] = `
{
  "canFailStop": false,
  "canFix": false,
  "canForce": false,
  "logLevel": "7-error",
  "willInit": false,
  "willShowHelp": false,
  "willShowVersion": false,
}
`;

exports[`getOptions B non existing target 1`] = `
{
  "canFailStop": false,
  "canFix": true,
  "canForce": false,
  "logLevel": "7-error",
  "willInit": false,
  "willShowHelp": false,
  "willShowVersion": false,
}
`;

exports[`getOptions C show help 1`] = `
{
  "canFailStop": false,
  "canFix": false,
  "canForce": false,
  "logLevel": "7-error",
  "willInit": false,
  "willShowHelp": true,
  "willShowVersion": false,
}
`;
